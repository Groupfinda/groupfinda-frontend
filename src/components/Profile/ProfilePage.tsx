import React from "react";
import {
  ScrollView,
  View,
  YellowBox,
  TouchableOpacity,
  AsyncStorage,
} from "react-native";
import {
  Avatar,
  Button,
  StyleService,
  Text,
  useStyleSheet,
  IconElement,
  useTheme,
  Icon,
  Layout,
  Divider,
  ListItem,
} from "@ui-kitten/components";
import { ImageOverlay } from "./extra/image-overlay.component";
import { ProfileSocial } from "./extra/profile-social.component";
import { DrawerGroupUser } from "./extra/drawer.component";
import { useNavigation } from "@react-navigation/native";
import { useApolloClient, useQuery } from "@apollo/react-hooks";
import { USER } from "../../graphql/queries";
import { Loading } from "../common";
import { SettingsIcon } from "./extra/icons";
import { BasicEventType } from "../types";

YellowBox.ignoreWarnings([
  "VirtualizedLists should never be nested inside plain ScrollViews",
]);

const PinIcon = (): IconElement => {
  const theme = useTheme();
  return (
    <Icon
      width={16}
      height={16}
      fill={theme["text-control-color"]}
      name="pin"
    />
  );
};

const QuestionsIcon = (): IconElement => {
  return <Icon width={16} height={16} fill="white" name="clipboard" />;
};

export default (): React.ReactElement => {
  const client = useApolloClient();
  const styles = useStyleSheet(themedStyle);
  const theme = useTheme();
  const navigation = useNavigation();

  const { loading, error, data } = useQuery(USER);

  if (error) {
    return (
      <View style={[styles.errorContainer]}>
        <View>
          <Text style={{ textAlign: "center" }}>
            Something went wrong: Please restart the application or contact the
            Development team
          </Text>
        </View>
      </View>
    );
  }

  if (loading || !data || !data.me) {
    return (
      <ScrollView style={styles.container}>
        <Loading visible={loading} />
      </ScrollView>
    );
  } else {
    return (
      <ScrollView style={styles.container}>
        <React.Fragment>
          <ImageOverlay
            style={styles.header}
            source={require("./temp/image-background.jpg")}
          >
            <Layout style={styles.layoutContainer}>
              <Layout style={styles.layout} level="1">
                <Button
                  status="control"
                  appearance="ghost"
                  accessoryLeft={QuestionsIcon}
                  onPress={() => {
                    navigation.navigate("Questions");
                  }}
                >
                  Questions
                </Button>
                <TouchableOpacity>
                  <Button
                    status="control"
                    appearance="ghost"
                    accessoryLeft={SettingsIcon}
                    onPress={() => {
                      navigation.navigate("ProfileSettings");
                    }}
                  >
                    Profile Settings
                  </Button>
                </TouchableOpacity>
              </Layout>
            </Layout>
            <Avatar
              style={{ width: 148, height: 148, marginBottom: 16 }}
              source={{ uri: data.me.avatar }}
            />
            <Text style={styles.profileName} category="h5" status="control">
              {data.me.firstName} {data.me.lastName}
            </Text>
            <View style={styles.locationContainer}>
              <PinIcon />
              <Text style={styles.location} status="control">
                {data.me.location}
              </Text>
            </View>
            <View style={styles.socialsContainer}>
              <ProfileSocial
                style={styles.profileSocial}
                hint="Events"
                value={data.me.profile.eventsRegistered.length}
              />
              <ProfileSocial
                style={styles.profileSocial}
                hint="Like/Dislikes"
                value={
                  data.me.profile.eventsLiked.length +
                  data.me.profile.eventsDisliked.length
                }
              />
              {/* <ProfileSocial
                style={styles.profileSocial}
                hint="Groups"
                value={data.me.groups.length}
              /> */}
              <ProfileSocial
                style={styles.profileSocial}
                hint="Interests"
                value={data.me.profile.userHobbies.length}
              />
            </View>
          </ImageOverlay>
          <Divider />
          <Layout style={{ padding: 12 }}>
            <Text style={{ fontWeight: "bold" }} category="h6">
              Profile Overview
            </Text>
            <ListItem
              disabled
              title="Hobbies/Interests"
              description={data.me.profile.userHobbies.join(", ")}
              accessoryLeft={() => (
                <Icon
                  width={20}
                  height={20}
                  fill={theme["color-primary-default"]}
                  name="brush-outline"
                />
              )}
            />
            <ListItem
              disabled
              title="Faculty"
              description={data.me.profile.userFaculty}
              accessoryLeft={() => (
                <Icon
                  width={20}
                  height={20}
                  fill={theme["color-primary-default"]}
                  name="book-outline"
                />
              )}
            />
            <ListItem
              disabled
              title="Year of Study"
              description={data.me.profile.userYearOfStudy}
              accessoryLeft={() => (
                <Icon
                  width={20}
                  height={20}
                  fill={theme["color-primary-default"]}
                  name="clock-outline"
                />
              )}
            />
          </Layout>
          <Divider />
          <DrawerGroupUser
            eventsLiked={data.me.profile.eventsLiked}
            eventsRegistered={data.me.profile.eventsRegistered.filter(
              (event: BasicEventType) =>
                event.dateOfEvent > new Date().getTime()
            )}
          />
          <Button
            status="danger"
            onPress={async () => {
              await AsyncStorage.removeItem("userToken");
              client.resetStore();
            }}
          >
            Log out
          </Button>
        </React.Fragment>
      </ScrollView>
    );
  }
};

const themedStyle = StyleService.create({
  errorContainer: {
    flex: 1,
    flexDirection: "column",
    alignItems: "stretch",
    justifyContent: "center",
    alignContent: "center",
    paddingHorizontal: 12,
  },
  layoutContainer: {
    flex: 1,
    flexDirection: "row",
    backgroundColor: "transparent",
  },
  layout: {
    width: "100%",
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    backgroundColor: "transparent",
  },
  container: {
    flex: 1,
    backgroundColor: "background-basic-color-2",
  },
  header: {
    paddingVertical: 24,
    alignItems: "center",
  },
  profileName: {
    zIndex: 1,
  },
  locationContainer: {
    flexDirection: "row",
    alignItems: "center",
  },
  location: {
    marginVertical: 8,
  },
  profileButtonsContainer: {
    flexDirection: "row",
    marginVertical: 32,
    marginHorizontal: 20,
  },
  profileButton: {
    flex: 1,
    marginHorizontal: 4,
  },
  socialsContainer: {
    flexDirection: "row",
    width: "75%",
    marginVertical: 8,
  },
  profileSocial: {
    flex: 1,
  },
});
