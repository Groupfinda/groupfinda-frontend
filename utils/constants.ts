export const eventCategories = [
  "Outdoors",
  "Adventure",
  "Technology",
  "Health",
  "Sports",
  "Fitness",
  "Learning",
  "Food & Drink",
  "Culture",
  "Music",
  "Social",
  "Games",
  "Arts",
  "Pets",
  "Hobbies",
  "Volunteer",
];
