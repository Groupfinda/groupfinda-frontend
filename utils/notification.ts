import Constants from 'expo-constants'
import * as Permissions from 'expo-permissions'
import { Platform } from "react-native";
import * as Notifications from 'expo-notifications'

export async function registerForPushNotificationsAsync() {
    let token;
    if (Constants.isDevice) {
        const { status: existingStatus } = await Permissions.getAsync(Permissions.NOTIFICATIONS);
        let finalStatus = existingStatus;
        if (existingStatus !== 'granted') {
            const { status } = await Permissions.askAsync(Permissions.NOTIFICATIONS);
            finalStatus = status;
        }

        if (finalStatus !== 'granted') {
            alert("Notifications allow us to send you updates on group matches and chat messages!")
            return
        }

        token = (await Notifications.getExpoPushTokenAsync()).data
        console.log(token)
    }

    if (Platform.OS === 'android') {
        Notifications.setNotificationChannelAsync('Groupfinda', {
            name: 'Groupfinda',
            importance: Notifications.AndroidImportance.MAX,
            vibrationPattern: [0, 250, 250, 250],
            lightColor: "#FF231F7C"
        })
    }

    return token
}

export type GroupfindaBody = {
    id: string;
    type: string;
}

export type AndroidNotificationData = GroupfindaBody
export type IOSNotificationData = {
    aps: {
        alert: {
            body: string;
            "launch-image": string;
            subtitle: string;
            title: string;
        }
        category: string;
        sound: string;
        "thread-id": string;
    }
    body: GroupfindaBody
}