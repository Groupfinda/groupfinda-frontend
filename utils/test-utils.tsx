import React from "react";
import { ApplicationProvider, IconRegistry } from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import { myTheme } from "../custom-theme";
import { EvaIconsPack } from "@ui-kitten/eva-icons";
import { MockedProvider, MockedResponse } from "@apollo/react-testing";
import { render, RenderAPI, RenderOptions } from "react-native-testing-library";

const AllProviders = (apolloMocks: MockedResponse[]) => ({
  children,
}: {
  children: JSX.Element;
}) => {
  return (
    <>
      <IconRegistry icons={EvaIconsPack} />
      <ApplicationProvider {...eva} theme={{ ...eva.light, ...myTheme }}>
        <MockedProvider mocks={apolloMocks} addTypename={false}>
          {children}
        </MockedProvider>
      </ApplicationProvider>
    </>
  );
};

type CustomRenderType = (
  component: React.ReactElement<any>,
  options?: RenderOptions,
  apolloProps?: MockedResponse[]
) => RenderAPI;

const customRender: CustomRenderType = (
  ui,
  options,
  apolloProps: MockedResponse[] = []
) => render(ui, { wrapper: AllProviders(apolloProps), ...options });

export * from "react-native-testing-library";
export { customRender as render };
