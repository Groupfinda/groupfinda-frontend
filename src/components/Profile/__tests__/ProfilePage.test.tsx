import React from "react";
import { render, cleanup } from "../../../../utils/test-utils"
import ProfilePage from "../ProfilePage"

jest.mock("@react-navigation/native", () => ({
    useNavigation: jest.fn(),
}));

afterEach(cleanup);

test("renders UI correctly", () => {
    const tree = render(<ProfilePage />);
    expect(tree).toMatchSnapshot();
})