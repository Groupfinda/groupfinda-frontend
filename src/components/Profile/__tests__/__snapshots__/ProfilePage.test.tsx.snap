// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders UI correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    style={
      Object {
        "backgroundColor": "#F7F9FC",
        "flex": 1,
      }
    }
  >
    <View>
      <Modal
        hardwareAccelerated={false}
        transparent={true}
        visible={false}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(0,0,0,0.5)",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <View
            style={
              Size {
                "height": 24,
                "width": 24,
              }
            }
          >
            <View
              style={
                Object {
                  "bottom": 0,
                  "height": 24,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    undefined,
                    Size {
                      "height": 12,
                      "width": 24,
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "rotate": "-165deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Size {
                          "height": 12,
                          "width": 24,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderColor": "#008EED",
                            "borderRadius": 12,
                            "borderWidth": 3.4,
                            "height": 24,
                            "width": 24,
                          },
                          undefined,
                        ]
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "bottom": 0,
                  "height": 24,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Object {
                      "top": 12,
                    },
                    Size {
                      "height": 12,
                      "width": 24,
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "translateY": -12,
                        },
                        Object {
                          "rotate": "345deg",
                        },
                      ],
                      "width": 24,
                    }
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Size {
                          "height": 12,
                          "width": 24,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderColor": "#008EED",
                            "borderRadius": 12,
                            "borderWidth": 3.4,
                            "height": 24,
                            "width": 24,
                          },
                          undefined,
                        ]
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </Modal>
    </View>
  </RCTScrollView>
</View>
`;
